src/core/net/tcp6.cpp

```diff
@@ -37,28 +37,49 @@
 
 #include "tcp6.hpp"
 
+#include <new>
+
 #include "common/code_utils.hpp"
 #include "common/error.hpp"
+#include "common/instance.hpp"
+#include "common/logging.hpp"
+#include "net/ip6.hpp"
 
 namespace ot {
 namespace Ip6 {
 
 Tcp::Tcp(Instance &aInstance)
     : InstanceLocator(aInstance)
+    , mTimer(aInstance, Tcp::HandleTimer)
+    , mFiringTimers(false)
+    , mEphemeralPort(kDynamicPortMin)
 {
+    OT_UNUSED_VARIABLE(mEphemeralPort);
 }
 
 Error Tcp::Endpoint::Initialize(Instance &aInstance, otTcpEndpointInitializeArgs &aArgs)
 {
-    OT_UNUSED_VARIABLE(aInstance);
-    OT_UNUSED_VARIABLE(aArgs);
+    Error error = kErrorNone;
 
-    return kErrorNotImplemented;
+    mContext                  = aArgs.mContext;
+    mEstablishedCallback      = aArgs.mEstablishedCallback;
+    mSendDoneCallback         = aArgs.mSendDoneCallback;
+    mSendReadyCallback        = aArgs.mSendReadyCallback;
+    mReceiveAvailableCallback = aArgs.mReceiveAvailableCallback;
+    mDisconnectedCallback     = aArgs.mDisconnectedCallback;
+
+    mInstance = &aInstance;
+
+    error = aInstance.Get<Tcp>().mEndpoints.Add(*this);
```

Comment:
Can this fail?

Reply:
No, it should never fail. It would only fail if this endpoint were already part of the linked list, which should never happen.

Reply:
Actually, I looked into this more, and I realized that `Add` actually checks if the element is already in the linked list. So I'm going to change this to do this check first (to protected against double-initialization) and exist if it's already part of the linked list.